lib/chef/resource/chef_client_config.rb

```diff
@@ -0,0 +1,307 @@
+
+      ```ruby
+      chef_client_config 'Create client.rb' do
+        chef_server_url 'https://chef.example.dmz'
+        additional_config <<~CONFIG
+          # Extra config code to safely load a gem into the client run.
+          # Since the config is Ruby you can run any Ruby code you want via the client.rb.
+          # It's a great way to break things, so be careful
+          begin
+            require 'aws-sdk'
+          rescue LoadError
+            Chef::Log.warn "Failed to load aws-sdk."
+          end
+        CONFIG
+      end
+      ```
+
+      **Setup two report handlers in the client.rb**:
+
+      ```ruby
+      chef_client_config 'Create client.rb' do
+        chef_server_url 'https://chef.example.dmz'
+        report_handlers [
+          {
+           'class' => 'ReportHandler1Class',
+           'arguments' => ["'FirstArgument'", "'SecondArgument'"],
+          },
+          {
+           'class' => 'ReportHandler2Class',
+           'arguments' => ["'FirstArgument'", "'SecondArgument'"],
+          },
+        ]
+      end
+      ```
+      DOC
+
+      # @todo policy_file or policy_group being set requires the other to be set so enforce that.
+      # @todo all properties for automate report
+      # @todo add all descriptions
+      # @todo validate handler hash structure
+
+      #
+      # @param [String, Symbol] prop_val the value from the property
+      #
+      # @return [Symbol] The symbol form of the symbol-like string, string, or symbol value
+      #
+      def string_to_symbol(prop_val)
+        if prop_val.is_a?(String) && prop_val.start_with?(":")
+          prop_val[1..-1].to_sym
+        else
+          prop_val.to_sym
+        end
+      end
+
+      property :config_directory, String,
+        description: "The directory to store the client.rb in.",
+        default: ChefConfig::Config.etc_chef_dir
+
+      property :user, String,
+        description: "The user that should own the client.rb file and the configuration directory if it needs to be created."
+
+      property :node_name, String,
+        description: "The name of the node. This determines which configuration should be applied and sets the `client_name`, which is the name used when authenticating to a #{ChefUtils::Dist::Server::PRODUCT}. If this value is not provided #{ChefUtils::Dist::Infra::PRODUCT} will use the node's FQDN as the node name. Leaving this setting blank and letting the client assign the FQDN of the node as the node_name during each #{ChefUtils::Dist::Infra::PRODUCT} run is recommended."
```

Comment:
It is actually probably better to set this to the node.name value, since otherwise the construction of the node name from the FQDN can fail due to lack of DNS.  The value in the config file isn't brittle like that, and shouldn't change.

Reply:
I had a default of node.name originally, but decided to go with the default out of config. I guess it does make sense if DNS is ever flakey to hard code that value.